### DISCUSSION TITLE
RoleHierarchy error

### DISCUSSION THREAD

**Discussion by julienmiclo** [symfony/symfony#48728] [Q&A] [answered]
Hello,
someone can help me ? I have a strange error.
This one :
`llegal offset type in isset or empty in /var/www/vendor/symfony/security-core/Role/RoleHierarchy.php line 43`.
```
public function getReachableRoleNames(array $roles): array    {
$reachableRoles = $roles;
foreach ($roles as $role) {
if (!isset($this->map[$role])) {
continue;
}
```
I am using User/Role ManyToMany relation, both are saved in database.
When i call getRoles in my User entity i do `return $this->roles->toArray();`

If you have docs or blogs for using Role in a database, i will take it.

Thank you.

**Comment by ahundiak** [comment]
Have you confirmed that $role is actually a string?  I'm guessing that it might be a Role entity?

**Comment by julienmiclo** [comment]
@ahundiak yes it's a array with Role entity. I save Role in database. I don't know how to work with. 
If i have a array string :  error when i make my jwt token,
and if i use array entity : error when i save my user.
Do you have docs, blog or other stuff for using entity Role with symfony ?

Thank you

**Comment by ahundiak** [comment, marked as answer]
For Symfony's authorization system, roles are strings.  End of story.  UserInterface::getRoles is required to return an array of strings.

If you want to use entities to contain roles then fine but it is still up to you to follow Symfony's method signatures.  This partly explains why most out of the box code simply stores roles as an array in the user table.  You clearly have some reason for using a many to many relation but once again you need to adjust getRoles accordingly.  Or just do what pretty much everyone else seems to do.

Fun fact.  Long long time ago (when Symfony 2 was initially released) there actually was a [Role class](https://github.com/symfony/symfony/blob/2.3/src/Symfony/Component/Security/Core/Role/Role.php) with a string property for the role itself.  Frankly it was a real PITA to use.  Seems like you were constantly converting between string and object.  I think everyone was pretty happy when it went away.

**Comment by avtarsm** [comment]
Hi, the other option is to store user Role in Json format in your user table.

**Comment by julienmiclo** [comment]
I found a solution, create a new property with an other name $dbRole and i use the getRoles for return string array from $dbRoles
